Add methods to QubitSparsePauli, QubitSparsePauliList, and PauliLindbladMap

DanPuzzuoli opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This is the third and final PR in a series to introduce `QubitSparsePauli`, `QubitSparsePauliList`, and `PauliLindbladMap`. It depends on #14383, which itself depends on #14283. The first two PRs establish the classes and their data storage, and this PR adds functionality. 

This PR introduces the following methods:

- `QubitSparsePauli`
   - `compose(other: QubitSparsePauli) -> QubitSparsePauli`: phase-less composition of two qubit-sparse Paulis.
   - `commutes(other: QubitSparsePauli) -> bool`: Returns `True` if `self` commutes with `other`, and `False` otherwise.
- `QubitSparsePauliList`
   - `apply_layout`: This is essentially a direct copy of the `SparseObservable.apply_layout` method with minor modifications to allow qubit label remappings.
- `PauliLindbladMap`
   - `scale_rates(scale_factor: float) -> PauliLindbladMap`: Return a new `PauliLindbladMap` with the rates scaled by the given factor (will be useful for noise amplification).
   - `inverse() -> PauliLindbladMap`: Return the mathematical inverse of the map.
   - `pauli_fidelity(pauli: QubitSparsePauli) -> float`: Return the Pauli fidelity of a given Pauli operator.
   - `compose(other: PauliLindbladMap) -> PauliLindbladMap`: Mathematical composition of `self` with `other`. This is implemented via concatenation of the underlying arrays representing the maps.
   - `drop_paulis(indices: List[int]) -> PauliLindbladMap`: Return a new map where the Paulis on the qubits `indices` are set to identity (essentially de-noising these qubits, when the map is viewed as an error map).
   - `keep_paulis(indices: List[int]) -> PauliLindbladMap`: Same as `drop_paulis`, but where the Paulis are dropped on the qubits not in `indices`.
   - `simplify(tol: float) -> PauliLindbladMap`: Simplify the map by collecting generator terms with the same operator (summing the rates), dropping terms whose rates are within `tol` of `0.0`, and dropping any terms where the operator is proportional to the identity (the corresponding generator of which is actually `0`). Aside from additionally dropping terms proportional to the identity, this is a copy of the corresponding `SparseObservable` method.
   - `apply_layout`: Utilizes `QubitSparsePauliList.apply_layout` to achieve the same goal.
   - `sample(n_samples: int, seed: int) -> (np.array(bool), QubitSparsePauliList)`: Draw `n_samples` samples from the map. A single sample consists of a boolean value (which represents whether the sampling requires propagating a classical `+/-1` for PEC) and a corresponding `QubitSparsePauli`. The `n_samples` are returned as an array of booleans and a corresponding `QubitSparsePauliList`.

### Details and comments

Depends on #14383 which depends on #14283. Apologies that this PR is huge, but to minimize rebasing it felt like it would be easier to deal with another large PR rather than a ton of small ones. @SamFerracin and I worked together on this, and have already reviewed each method that the other wrote.

Additional notes:
- Can the sampling in `PauliLindbladMap.sample` be written in a way to enable parallelization? Each sample is currently drawn independently.

DanPuzzuoli committed:
Add methods to QubitSparsePauli, QubitSparsePauliList, and PauliLindbladMap

This is a roll-up of the previous done on top of the addition of the
`PauliLindbladMap` class to add extra methods to the named structs.

Co-authored-by: Samuele Ferracin <sam.ferracin@ibm.com>
Co-authored-by: Ian Hincks <ian.hincks@gmail.com>

ihincks commented on crates/accelerate/src/pauli_lindblad_map/pauli_lindblad_map_class.rs:
```diff
@@ -0,0 +1,1734 @@
+    ) -> PyResult<Bound<'py, PyTuple>> {
+        //PyResult<Bound<'py,(PyArray1<bool>, PyQubitSparsePauliList)>> {
+        let inner = self.inner.read().map_err(|_| InnerReadError)?;
+        let (signs, paulis) = inner.sample(n_samples, seed)?;
```
Since this method may be a workhorse in production, and since that environment may benefit from threading, I think this is a case where it will be worth it to give up the GIL.

```suggestion
        let (signs, paulis) = py.allow_threads(|| inner.sample(n_samples, seed))?;
```

To measure this I used this setup:

```python
from qiskit.quantum_info import PauliLindbladMap
from concurrent.futures import ThreadPoolExecutor

import numpy as np

def generate_random_pauli_list(
    num_terms: int,
    num_qubits: int,
    sparsity: int,
    coeff_range: tuple[float, float]
) -> list[tuple[str, float]]:
    paulis = ['X', 'Y', 'Z']
    result = []

    for _ in range(num_terms):
        chars = ['I'] * num_qubits

        indices = np.random.choice(num_qubits, sparsity, replace=False)

        for idx in indices:
            chars[idx] = np.random.choice(paulis)

        pauli_string = ''.join(chars)
        coeff = np.random.uniform(*coeff_range)
        result.append((pauli_string, coeff))

    return result


ops = [
    PauliLindbladMap.from_list(generate_random_pauli_list(600, 150, 4, [0, .1]))
    for _ in range(100)
]
```

and then ran this notebook cell built with the latest commit of this branch, and then again built with the code suggestion of this comment:

```python
%%timeit 
with ThreadPoolExecutor(12) as executor:
    results = list(executor.map(lambda lind, seed: lind.sample(1650, seed), ops, range(len(ops))))
```

The results are respectively `3.84 s ± 55.5 ms per loop` and `558 ms ± 32.3 ms`, about 6.5x

DanPuzzuoli replied:
Nice LGTM. Committing.

jakelishman replied:
Nice - this is the first real meaningful use of releasing the GIL in Qiskit!  So far we've not really bothered much because moves to Rust have typically been to take advantage of Rust-space multithreading in various algorithms, and we still generally needed to occasionally act on Python space.

At some point we'll want to sweep through and do more of this sort of handling in the future (though in many places, we'll need to make our thread-control story better, so users can dynamically select _where_ the threading gets applied).  The other place for considering Python interaction, which would be relevant here, but no need to change anything: we fairly reliably _don't_ check for Python signals during execution of Rust-space code, which means we typically just ignore `KeyboardInterrupt` signals until we happen to pass control back to Python space for any reason, which isn't really a great situation.

ihincks commented on crates/accelerate/src/pauli_lindblad_map/pauli_lindblad_map_class.rs:
```diff
@@ -0,0 +1,1734 @@
+        py: Python<'py>,
+        n_samples: u64,
+        seed: Option<u64>,
+    ) -> PyResult<Bound<'py, PyTuple>> {
```
Btw, while making my other comment, i noticed that the sample method, from python, doesn't seem to recognize that the seed is optional. I get `TypeError: PauliLindbladMap.sample() missing 1 required positional argument: 'seed'` if only specifying the size.

jakelishman replied:
Dan: you need to add a signature annotation to tell PyO3 about Python-space defaults, if you want them to exist.  Something like this:
```rust
#[pyo3(signature = (n_samples, *, seed=None))]
pub fn sample<'py>(...)
```

jakelishman commented:
I've done the same roll-up and force push to this branch as I did to #14383 (see https://github.com/Qiskit/qiskit/pull/14383#issuecomment-2922061237), just to facilitate me reviewing.  The previous head commit of this branch was ed58acca82e79ab54d2279277d7dcdb8d5aa6387 - if you have local changes that you need any hand in saving/reapplying, let me know and I can help.

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/pauli_lindblad_map_class.rs:
```diff
@@ -264,6 +284,208 @@ impl PauliLindbladMap {
+    pub unsafe fn scale_rates(self, scale_factor: f64) -> Self {
+        let new_rates = self.rates.iter().map(|r| scale_factor * r).collect();
+        unsafe { PauliLindbladMap::new_unchecked(new_rates, self.qubit_sparse_pauli_list.clone()) }
+    }
```
Writing `unsafe fn` means "the caller of this function has safety guarantees that must be upheld, which the compiler can't do".  You don't need a function to be `unsafe fn` to use an `unsafe {}` block inside it, and most of the time if you're using `unsafe`, you're providing stable _safe_ public functions that only use `unsafe` in ways that can be validated within the function as being sound.

This is one of the latter cases - `self` can't be in an invalid state from using safe code, and there's no value of `scale_factor` that can cause this to be unsafe (if there is a value that's forbidden, more typically you'd return `Result` or panic, depending on whether it's a recoverable error or not).  You can entirely drop the "safety" section of docs, and turn `pub unsafe fn` into just `pub fn`.

DanPuzzuoli replied:
I've completely dropped both usages of unsafe here as the `PauliLindbladMap::new_unchecked` method was removed in the preceding PR.

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/pauli_lindblad_map_class.rs:
```diff
@@ -264,6 +284,208 @@ impl PauliLindbladMap {
+    /// need to construct the map with the safe PauliLindbladMap::new.
+    pub unsafe fn inverse(self) -> Self {
+        unsafe { self.scale_rates(-1.) }
+    }
```
Same comment here about safety - this method is completely safe to call.

DanPuzzuoli replied:
Dropped the unsafes.

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/pauli_lindblad_map_class.rs:
```diff
@@ -264,6 +284,208 @@ impl PauliLindbladMap {
+                qubit_sparse_pauli_list,
+            ))
+        }
+    }
```
The algorithm for this `compose` has no problems being entirely in-place, so you may well want to have this mutate `self` - it makes it cheaper to build up the noise map, because in the current form, calling compose in a loop is potentially quadratic.

DanPuzzuoli replied:
I'm not totally sure but I don't believe composition needs to be super performant. I think an in-place version of composition could be implemented later.

jakelishman replied:
Ok, sure - this at least will match the behaviour of other `quantum_info` classes.  Usually I prefer to make the performant one the default wherever possible, so you don't have to use an `inplace=True` option manually, but I think consistency with other classes is a good choice too.

Fwiw, there's "doesn't need to be performant" and then there's "accidentally quadratic" haha

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/pauli_lindblad_map_class.rs:
```diff
@@ -264,6 +284,208 @@ impl PauliLindbladMap {
+        &self,
+        n_samples: u64,
+        seed: Option<u64>,
+    ) -> Result<(Vec<bool>, QubitSparsePauliList), CoherenceError> {
```
It's possible you'll want a `sample_into` method that takes a `&mut [bool]` and `&mut QubitSparsePauliList`, so you can re-use the allocation if you need to call this multiple times.  It's quite common in Rust to build up chains of abstractions like this, to allow callers to go in at the level of abstraction/reuse that makes sense for them.  It's a function of the language _not_ having keyword arguments, and being compiled so many functions can be inlined into their callsite, and you don't pay for them at all:

```rust
pub fn sample_into_with_rng(&self, &mut rng: RngCore, signs: &mut Vec<bool>, paulis: &mut QubitSparsePauliList) {
    let n_samples = signs.len();
    // ... most of the current method ...
    // note we take `&mut Vec<bool>` even though else where I suggested `&[bool]`
    // more typical, because here, you'll need to push to the allocation.
}
#[inline]
pub fn sample_into(&self, seed: Option<u64>, signs: &mut Vec<bool>, paulis: &mut QubitSparsePauliList) {
    let mut rng = match seed { ... };
    self.sample_into(&mut rng, signs, paulis)
}
pub fn sample(&self, num_samples: u64, seed: Option<u64>) -> (Vec<bool>, QubitSparsePauliList) {
    let mut signs = Vec::with_capacity(num_qubits);
    let mut paulis = QubitSparsePauliList::with_capacity(num_samples, <something>);
    self.sample_into(seed, &mut signs, &mut paulis);
    (signs, paulis)
}
```

You don't have to do this, or to include all three levels of abstraction that I did, but the idea of the `sample_into` method is quite possibly something that'll be useful to you in Rust space in the future.

We'd have to think more if we wanted to have the "signs" reuse be into an object that's managed by Python, though.

DanPuzzuoli replied:
I think it might make sense to include this in subsequent improvements focused on performance when we have more time to think about the interfaces.

jakelishman replied:
Yep, this is totally doable as a future API extension.

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/pauli_lindblad_map_class.rs:
```diff
@@ -264,6 +284,208 @@ impl PauliLindbladMap {
+        for _ in 0..n_samples {
+            let mut random_sign = true;
+            let mut random_pauli =
+                QubitSparsePauli::new(self.num_qubits(), Box::new([]), Box::new([]))?;
```
This should be infallible (and consequently so should the entire `sample` method) - you presumably have a `QubitSparsePauli::identity(self.num_qubits())` which shouldn't need coherence checks?

DanPuzzuoli replied:
Ah nice - yeah this method didn't exist, but I've just added it and replaced the above line with:
```
let mut random_pauli = QubitSparsePauli::identity(self.num_qubits());
```

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/pauli_lindblad_map_class.rs:
```diff
@@ -1069,6 +1293,299 @@ impl PyPauliLindbladMap {
+        &self,
+        py: Python<'py>,
+        n_samples: u64,
+        seed: Option<u64>,
```
Qiskit convention is usually `num_samples`, not `n_samples`.

DanPuzzuoli replied:
changed `n_samples` -> `num_samples`

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/pauli_lindblad_map_class.rs:
```diff
@@ -1069,6 +1293,299 @@ impl PyPauliLindbladMap {
+        py: Python<'py>,
+        n_samples: u64,
+        seed: Option<u64>,
+    ) -> PyResult<Bound<'py, PyQubitSparsePauliList>> {
```
Same here about `n_samples`.

DanPuzzuoli replied:
changed `n_samples` -> `num_samples`

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/pauli_lindblad_map_class.rs:
```diff
@@ -1069,6 +1293,299 @@ impl PyPauliLindbladMap {
+
+        let (_, paulis) = py.allow_threads(|| inner.sample(n_samples, seed))?;
+
+        paulis.into_pyobject(py)
```
This can be a follow-up: you might consider modifying the Rust methods so you never even allocate storage space for the signs if you don't need to sample them.

jakelishman reviewed (commented):
This review is just for the `PauliLindbladMap` struct, and I was racing a little bit towards the end already - I'll revisit on Monday.  I still need to look through `QubitSparsePauli` and the tests.

DanPuzzuoli committed:
rebase and tests passing

DanPuzzuoli committed:
adding QubitSparsePauli identity function

DanPuzzuoli committed:
changing n_samples -> num_samples

DanPuzzuoli committed:
formatting

DanPuzzuoli commented:
@jakelishman 

I've rebased off of the changes to the `PauliLindbladMap` PR and believe I've addressed your comments. As mentioned in one of the responses to your comments, I've added an `identity` constructor to `QubitSparsePauli`.

jakelishman committed:
Fix documentation build

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/pauli_lindblad_map_class.rs:
```diff
@@ -224,6 +244,188 @@ impl PauliLindbladMap {
+        }
+        new_boundaries.push(current_boundary - num_dropped_paulis);
+
+        Self::new(
+            self.rates().to_vec(),
+            QubitSparsePauliList::new(self.num_qubits(), new_paulis, new_indices, new_boundaries)
+                .unwrap(),
+        )
```
Strictly, I think here you can return
```rust
Self {
    qubit_sparse_pauli_list: QubitSparsePauliList::new(...).expect("all transformations retain coherence"),
    ..self.clone()
}
```
or something like that, and have the function be infallible - I don't _think_ there's any way you can lose coherence here?

Methods on a struct can always see all of the struct internals, so you can use the direct constructor form when it's convenient.

---

This isn't super important - it can always be done later, if it's appropriate.

DanPuzzuoli replied:
Yeah I need to learn all the correct ways to handle errors (or to ignore them when I know there is no way for there to be an error). There should be no way to lose coherence here unless there is a bug in the code of the method.

jakelishman replied:
Yeah no worries, this can totally just be interface pruning later.  We have tonnes of stuff like this all over the place in Qiskit - it's easy enough to tidy up later.

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/pauli_lindblad_map_class.rs:
```diff
@@ -224,6 +244,188 @@ impl PauliLindbladMap {
+            }
+        }
+
+        Ok(fid)
```
Maybe there's a speed-up available if you calculate the fidelity in logarithmic space and then only called `exp` at the very end?  I think the maths shakes out right.  Can easily be a follow-up.

DanPuzzuoli replied:
Yes of course, good catch - just changed this.

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/pauli_lindblad_map_class.rs:
```diff
@@ -224,6 +244,188 @@ impl PauliLindbladMap {
+        &self,
+        num_samples: u64,
+        seed: Option<u64>,
+    ) -> Result<(Vec<bool>, QubitSparsePauliList), CoherenceError> {
```
I don't think it's possible for this method to return the error state any more - I think you can change its signature just to the `(Vec<bool>, QubitSparsePauliList)` thing?

DanPuzzuoli replied:
Correct - I'm trying to remember when it ever had an error. The python-space method does check if there are any negative rates and will raise an error if there are, but this one maybe never did. 

Changed the type.

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/pauli_lindblad_map_class.rs:
```diff
@@ -224,6 +244,188 @@ impl PauliLindbladMap {
+            new_indices.extend_from_slice(indices);
+            new_boundaries.push(new_indices.len());
+        }
+        Self::new(
+            new_rates,
+            QubitSparsePauliList::new(self.num_qubits(), new_paulis, new_indices, new_boundaries)
+                .unwrap(),
+        )
+        .unwrap()
```
Similar comment to a previous one about how you can use your privileged access to the constructor to avoid the terminal `.unwrap()`.  When looking at performance in the future, we can also consider adding a `QubitSparsePauliList::new_unchecked` variant to avoid the result stuff in that too, in cases where we know it's safe.

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/pauli_lindblad_map_class.rs:
```diff
@@ -1010,6 +1214,308 @@ impl PyPauliLindbladMap {
+            )));
+        }
+
+        Ok(inner
+            .drop_paulis(
+                (0..self.num_qubits()?)
+                    .filter(|index| !indices.contains(index))
+                    .collect(),
+            )?
+            .into())
```
For a follow-up: perhaps a nicer interface will be for the Rust-space `drop_paulis` to be generic over some `FnMut(Qubit) -> bool`, then the `drop_paulis` and `keep_paulis` can both use the same mechanism - convert to a set, and it's just that one tests for the presence, and one tests for absence.

DanPuzzuoli replied:
Agreed!

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/pauli_lindblad_map_class.rs:
```diff
@@ -1023,6 +1529,58 @@ impl PyPauliLindbladMap {
+        Ok(result)
+    }
+
+    fn __matmul__<'py>(
+        &self,
+        other: &Bound<'py, PyAny>,
+    ) -> PyResult<Bound<'py, PyPauliLindbladMap>> {
+        self.compose(other)
+    }
```
Is `__matmul__` the correct operator for the ordering of `compose` here?  For operator-like objects, `a.compose(b)` corresponds in matrix algebra to `b @ a` (and `quantum_info` typically uses the `a & b` operator for it).

I _think_ that the `compose` here just becomes a sum over the generator terms, right, so it's commutative?  If so, then `__matmul__` is fine, though I do slightly worry about the feeling of consistency.

DanPuzzuoli replied:
My reasoning for using `__matmul__`: A `PauliLindbladMap` instance ultimately represents a linear map on matrices, and in general, composition of these is a form of matrix multiplication. It just so happens that the set of maps `PauliLindbladMap` is representing is pair-wise commutative, and due to the parameterization in terms of an exponentiated generator, composition can be implemented by summing the generators.

So, in my opinion, `__matmul__` makes sense to me, but I'm also open to dropping this if you're not comfortable with it or if there are inconsistencies with other parts of `qiskit` (e.g. if commutative composition is always represented by `__mul__`).

jakelishman replied:
I agree with the association with matrix multiplication, it's mostly just the implied argument ordering I'm questioning - `a.compose(b)` means `b @ a` normally in Qiskit: it's the operator tha "applies 'a' and then applies 'b'".

jakelishman replied:
If it's commutative, then I'm not bothered.

DanPuzzuoli replied:
Ah true true - sorry misread.

jakelishman reviewed (commented):
Thanks Dan - this looks sensible to me.

I reviewed mostly for the _interfaces_ the PR introduces - I didn't go into any great depth checking that the implementations were mathematically correct - largely I trust that your tests caught anything like that (and most of the mathematics seems straightforwards).

jakelishman committed:
Fix incorrect method name

jakelishman committed:
Fix clippy complaint

jakelishman commented:
Sorry, I clearly missed a couple of method renames following changes in the other PR - should be as straightforwards as deleting the `get_` prefixes.

DanPuzzuoli committed:
responding to comments

jakelishman reviewed (approved):
Ok, I think this is sound as-is now, thanks so much for all the work, Dan (and Sam, and Ian)!

Everything left we've been talking about can be modified, even within backwards-compatibility guarantees, so I'm not concerned.
